### PR TITLE
Tacho : memory-management for SpMV option

### PR DESCRIPTION
**Pull request by iyamazaki**
@trilinos/shylu 

## Motivation

This PR replaces the views used to store CRS with subviews.

## Testing

Tested on V100 and Mi250.

## Additional Information

Tacho uses a view of structures (one structure per level). Previously, we used views in the structure to store CRS, which could have lead to memory leak.

### CODE REVIEW

File: packages/shylu/shylu_node/tacho/src/impl/Tacho_NumericTools_LevelSet.hpp [hunk 1]
```diff
@@ -2235,125 +2383,125 @@ class NumericToolsLevelSet : public NumericToolsBase<ValueType, DeviceType> {
...
                         CUSPARSE_INDEX_BASE_ZERO, computeType);
     }
     // Call SpMV/SPMM
     cusparseStatus_t status;
+    cusparseOperation_t opL = (s0.spmv_explicit_transpose ? CUSPARSE_OPERATION_NON_TRANSPOSE : CUSPARSE_OPERATION_TRANSPOSE);
     if (nrhs > 1) {
       if (lvl == nlvls-1) {
+        // start : destroy previous
+        cusparseDestroyDnMat(matL);
         // start : create DnMat for T
```
Comment (ndellingwood) [inline packages/shylu/shylu_node/tacho/src/impl/Tacho_NumericTools_LevelSet.hpp:2428]: Minor typo: 
```suggestion
        // start : create DnMat for L
```

File: packages/shylu/shylu_node/tacho/src/impl/Tacho_NumericTools_LevelSet.hpp [hunk 2]
```diff
@@ -2235,125 +2383,125 @@ class NumericToolsLevelSet : public NumericToolsBase<ValueType, DeviceType> {
...
+      status = cusparseSpMM(cusparseHandle, opL, CUSPARSE_OPERATION_NON_TRANSPOSE,
+                            &alpha, s0.L_cusparse, 
+                                    matX,
+                            &beta,  matY,
+                            computeType, TACHO_CUSPARSE_SPMM_ALG, (void*)buffer_L.data());
     } else {
       if (lvl == nlvls-1) {
+        // start : destroy previous
+        cusparseDestroyDnVec(vecL);
         // start : create DnMat for T
```
Comment (ndellingwood) [inline packages/shylu/shylu_node/tacho/src/impl/Tacho_NumericTools_LevelSet.hpp:2444]: Minor typo: 
```suggestion
        // start : create DnMat for L
```

**Review by ndellingwood — APPROVED** [review]
Just a couple minor typos in comments marked, not necessary to retest but worth updating in follow-up PR. Changes looked reasonable in the refactor
@iyamazaki can you confirm testing of the rocsparse changes in a Hip build and they passed tests?

**Comment by iyamazaki** [comment]
Thank you, @ndellingwood. Yes, I tested on Mi250 (Frontier).

**Comment by ndellingwood** [comment]
@iyamazaki thanks for confirming, looks like you're ready to merge :)

**Pull request merged**